crates/bounded_planet_core/src/networking/systems.rs

```diff
@@ -173,103 +181,107 @@ pub async fn handle_connection(
     }
 }
 
-/// Pull packets from an mpsc and send them to the given stream
-async fn send_to_stream(
-    connection_id: ConnectionId,
-    stream_id: StreamId,
-    send: SendStream<TlsSession>,
-    mut recv: UnboundedReceiver<Arc<Packet>>,
-    event_sender: UnboundedSender<ReceiveEvent>
-) {
-    let mut send = BoundedPlanetSendStream::new(send);
-    while let Some(pkt) = recv.recv().await {
-        if let Err(err) = send.send_packet(&pkt).await {
-            event_sender.send(ReceiveEvent::NetworkError(
-                NetworkError::SendError {
-                    connection_id,
-                    stream_id,
-                    err,
+async fn send_to_streams(connection_id: ConnectionId, mut conn: quinn::Connection, mut recv: UnboundedReceiver<(StreamType, Arc<Packet>)>, event_sender: UnboundedSender<ReceiveEvent>) {
+
+    // Create a list of open streams. When a request comes in to send over a non-existant stream open it and store it here.
+    // Streams are never closed. This is fine since there are a fixed number fo streams (as defined in the StreamType enum).
+    let mut stream_lookup = Vec::new();
+
+    loop {
+        match recv.recv().await {
+
+            // Once the receiver receives `None` that indicates all senders have been dropped. This loop can safely end because
+            // it's impossible for any more messages to arrive at this MPSC receiver.
+            None => { break; }
+
+            // Send a packet through a stream
+            Some((stream_type, pkt)) => {
+                // Find (or create) the sender for this stream
+                let sender = match get_stream_sender(&stream_type, &mut stream_lookup, &mut conn).await {
+                    Ok(sender) => sender,
+                    Err(err) => {
+                        event_sender.send(ReceiveEvent::NetworkError(
+                             NetworkError::ConnectionError(err)
+                        )).expect("Failed to send error event!");
+                        break;
+                    },
+                };
+
+                // Send the packet through the stream
+                if let Err(err) = sender.send_packet(&pkt).await {
+                    event_sender.send(ReceiveEvent::NetworkError(
+                        NetworkError::SendError {
+                            connection: connection_id,
+                            stream: stream_type,
+                            err,
+                        }
+                    )).expect("Failed to send error event!");
                 }
-            )).expect("Failed to send error event!");
+            }
         }
     }
 }
+
+async fn get_stream_sender<'a>(
+    stream_type: &StreamType,
+    stream_lookup: &'a mut Vec<BoundedPlanetSendStream<TlsSession>>,
+    conn: &mut quinn::Connection
+) -> Result<&'a mut BoundedPlanetSendStream<TlsSession>, ConnectionError>
+{
+    // Calculate the index of this sender simply as the enum variant index
+    let idx = *stream_type as usize;
```

Comment:
I'm not sure it's guaranteed that for an enum with N variants, the tag values are 0 .. N-1. This may cause an issue of allocating way too many entries. That said, the tag is capped at u8, and 256 isn't really that many.

Reply:
From what I've read it does seem to be guaranteed. Although I can't find a solid reference that says one way or the other.